x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/oauth2/RestGetTokenAction.java

```diff
@@ -124,6 +129,12 @@ public void onFailure(Exception e) {
                     ((ElasticsearchSecurityException) e).getHeader("error_description").size() == 1) {
                 sendTokenErrorResponse(TokenRequestError.INVALID_GRANT,
                         ((ElasticsearchSecurityException) e).getHeader("error_description").get(0), e);
+            } else if (e instanceof ElasticsearchSecurityException
+                    && "failed to authenticate user, gss context negotiation not complete".equals(e.getMessage())
+                    && ((ElasticsearchSecurityException) e).getHeader(KerberosAuthenticationToken.WWW_AUTHENTICATE) != null
+                    && ((ElasticsearchSecurityException) e).getHeader(KerberosAuthenticationToken.WWW_AUTHENTICATE).size() == 1) {
+                sendTokenErrorResponse(TokenRequestError.UNAUTHORIZED_CLIENT,
```

Comment:
I'm torn between using an error code that makes no sense in this case and defining a custom one. I think I'd prefer us to define a new error code (that is not in RFC6749) ( for the same arguments we didn't want to (ab)use client_credentials grant for this)